src/api/block_builder/input_selection/sender_issuer.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! sender and issuer features input selection
+
+use std::collections::HashSet;
+
+use bee_block::{
+    address::Address,
+    output::{dto::OutputDto, feature::Features, AliasOutput, NftOutput, Output},
+};
+use crypto::keys::slip10::Chain;
+
+use crate::{
+    api::{
+        address::search_address,
+        input_selection::{helpers::is_basic_output_address_unlockable, output_contains_address},
+        ClientBlockBuilder,
+    },
+    constants::HD_WALLET_TYPE,
+    secret::types::{InputSigningData, OutputMetadata},
+    Error, Result,
+};
+
+impl<'a> ClientBlockBuilder<'a> {
+    pub(crate) async fn get_inputs_for_sender_and_issuer(
+        &self,
+        utxo_chain_inputs: &[InputSigningData],
+    ) -> Result<(bool, Vec<InputSigningData>)> {
+        log::debug!("[get_inputs_for_sender_and_issuer]");
+        let mut force_use_all_inputs = false;
+        let mut required_inputs = Vec::new();
+        let bech32_hrp = self.client.get_bech32_hrp().await?;
+        let current_time = self.client.get_time_checked().await?;
+
+        let mut all_required_addresses = HashSet::new();
+        for output in &self.outputs {
+            let mut required_addresses = HashSet::new();
```

Comment:
Can we not just use `all_required_addresses` and potentially remove from it if necessary? We are double-allocating otherwise

Reply:
But then we would iterate over all addresses from all outputs every time in L56 instead of only the 1 or 2 addresses from this output

Reply:
Maybe you don't even need to iterate at all if you make it an inline function and call it twice idk, feel free to keep as it is now though, just suggesting